zkevm-circuits/src/evm_circuit/util/constraint_builder.rs

```diff
@@ -0,0 +1,555 @@
+        Vec<Preset<F>>,
+    ) {
+        let mut constraints = self.constraints;
+        let mut presets = Vec::new();
+
+        for (row, usage) in self.curr.rows.iter().zip(self.row_usages.iter()) {
+            if usage.is_byte_lookup_enabled {
+                constraints.push((
+                    "Enable byte lookup",
+                    row.qs_byte_lookup.expr() - 1.expr(),
+                ));
+            }
+
+            presets.extend(
+                row.cells[usage.next_idx..]
+                    .iter()
+                    .map(|cell| (cell.clone(), F::zero())),
+            );
+            presets.push((
+                row.qs_byte_lookup.clone(),
+                if usage.is_byte_lookup_enabled {
+                    F::one()
+                } else {
+                    F::zero()
+                },
+            ));
+        }
+
+        let execution_state_selector =
+            self.curr.execution_state_selector(self.execution_state);
+
+        (
+            constraints
+                .into_iter()
+                .map(|(name, constraint)| {
+                    (name, execution_state_selector.clone() * constraint)
+                })
+                .collect(),
+            self.lookups
+                .into_iter()
+                .map(|lookup| {
+                    lookup.conditional(execution_state_selector.clone())
+                })
+                .collect(),
+            presets,
+        )
+    }
+
+    pub(crate) fn randomness(&self) -> Expression<F> {
+        self.randomness.clone()
+    }
+
+    pub(crate) fn execution_state(&self) -> ExecutionState {
+        self.execution_state
+    }
+
+    pub(crate) fn rw_counter_offset(&self) -> usize {
+        self.rw_counter_offset
+    }
+
+    pub(crate) fn program_counter_offset(&self) -> usize {
+        self.program_counter_offset
+    }
+
+    pub(crate) fn stack_pointer_offset(&self) -> i32 {
+        self.stack_pointer_offset
+    }
+
+    // Query
+
+    pub(crate) fn query_bool(&mut self) -> Cell<F> {
+        let [cell] = self.query_cells::<1>(false);
+        self.require_boolean("Constrain cell to be a bool", cell.expr());
+        cell
+    }
+
+    pub(crate) fn query_byte(&mut self) -> Cell<F> {
+        let [cell] = self.query_cells::<1>(true);
+        cell
+    }
+
+    pub(crate) fn query_cell(&mut self) -> Cell<F> {
+        let [cell] = self.query_cells::<1>(false);
+        cell
+    }
+
+    pub(crate) fn query_word(&mut self) -> Word<F> {
+        Word::new(self.query_bytes(), self.randomness.clone())
+    }
+
+    pub(crate) fn query_bytes<const N: usize>(&mut self) -> [Cell<F>; N] {
+        self.query_cells::<N>(true)
+    }
+
+    fn query_cells<const N: usize>(&mut self, is_byte: bool) -> [Cell<F>; N] {
+        let mut cells = Vec::with_capacity(N);
+
+        for (row, usage) in
+            self.curr.rows.iter().zip(self.row_usages.iter_mut())
+        {
+            if usage.is_byte_lookup_enabled != is_byte {
```

Comment:
The code here requires all byte cells to be in the same row. Is this too restricted? There could be the case that you only allocate one byte cell instead of a word.

Reply:
Yeah there will be the case, but when further bytes are queried, those unused cells will be returned by `ConstraintBuilder`, which would try to use all available cells queried (it panics when cells are not enough).

This is a small optimization to use an extra advice column to not blow up amount of byte lookups (that already exist in previously), let me document it now.